### PR TITLE
[GH239] Replace ALB ingress controller with AWS load balancer controller

### PR DESCRIPTION
**Pull request by paulbes**
Replace ALB ingress controller with AWS load balancer controller (https://github.com/kubernetes-sigs/aws-load-balancer-controller)

## Description
This is a continuation of the "due diligence" process we have going, where we are moving towards the latest versions of our dependencies. 

AWS Load Balancer Controller is a controller to help manage Elastic Load Balancers for a Kubernetes cluster.

- It satisfies Kubernetes Ingress resources by provisioning Application Load Balancers.
- It satisfies Kubernetes Service resources by provisioning Network Load Balancers.

## Motivation and Context
The motivation is well described in #239 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I have updated the release notes (for the next release).

### CODE REVIEW

File: pkg/api/core/service_serviceaccount.go
Comment (deifyed) [inline pkg/api/core/service_serviceaccount.go:186]: Why do we create a service account just to delete it (line 197)? Or does this function just create some kind of config despite its name

File: pkg/api/core/store/noop/helm_noop.go
Comment (deifyed) [inline pkg/api/core/store/noop/helm_noop.go:15]: Missing 'T' at the end of function name.

File: pkg/apis/okctl.io/v1alpha1/cluster_v1alpha1.go
Comment (deifyed) [inline pkg/apis/okctl.io/v1alpha1/cluster_v1alpha1.go:251]: Maybe we should get rid of the ALBIngressController? (Not in this PR) Or is it some kind of difference in functionality that we need to keep around

File: pkg/clusterconfig/clusterconfig.go
Comment (deifyed) [inline pkg/clusterconfig/clusterconfig.go:230]: Name is misleading. Should be something like `NewAWSLoadBalancerControllerServiceAccountConfig`

File: pkg/helm/charts/awslbc/awslbc.go
Comment (deifyed) [inline pkg/helm/charts/awslbc/awslbc.go:38]: Refactor before or after merge?

File: pkg/helm/charts/awslbc/awslbc.go
Comment (deifyed) [inline pkg/helm/charts/awslbc/awslbc.go:35]: What is this? Could we automate creating certificates for applications?

**Review by deifyed — CHANGES_REQUESTED** [review]
8 code smells and a couple of comments

Good work though

**Comment by yngvark** [comment]
@paulbes What do users with existing okctl clusters do to get this new AWS Load Balancer controller? I.e. how do they upgrade?

Should we write it in the user docs? And if it isn't possible to upgrade, still write something in the user docs aboutit.

**Comment by deifyed** [comment]
@yngvark With `apply cluster` people should simply be able to set the old one to false and the new one to true 😍

File: pkg/api/core/service_serviceaccount.go
Comment (paulbes) [reply pkg/api/core/service_serviceaccount.go:186]: I am not happy about this functionality, but essentially we are using `eksctl` to delete the service account for us. To do this, we need to provide a config the cluster name, region, service account name, etc.

If you look at the package name: `clusterconfig` you can see that your assumption is correct.

File: pkg/apis/okctl.io/v1alpha1/cluster_v1alpha1.go
Comment (paulbes) [reply pkg/apis/okctl.io/v1alpha1/cluster_v1alpha1.go:251]: Nope, we should remove it, just figured I would do it in a separate PR. 👍🏽

File: pkg/clusterconfig/clusterconfig.go
Comment (paulbes) [reply pkg/clusterconfig/clusterconfig.go:230]: Given the name of the package, I am not sure this is required? I even wonder if the linter might not give us a warning if we try to do that🤔


From: https://blog.golang.org/package-names
> Avoid stutter. Since client code uses the package name as a prefix when referring to the package contents, the names for those contents need not repeat the package name. The HTTP server provided by the http package is called Server, not HTTPServer. Client code refers to this type as http.Server, so there is no ambiguity.

File: pkg/helm/charts/awslbc/awslbc.go
Comment (paulbes) [reply pkg/helm/charts/awslbc/awslbc.go:36]: I believe it refers to this: https://github.com/jetstack/cert-manager, and yes, we could potentially use it now. It is essentially a service that uses lets encrypt to create certificates.

The main problem is that `AWS load balancer controller` and `cert-manager` kind of step on each others toes. With an ALB we typically terminate the encryption at the ALB and pass the request to the k8s pods in plaintext. Cert-manager attempts to terminate the encryption on the pod level. Now that we are using `AWS load balancer controller`, which supports Network Load Balancers (NLB). This might actually be possible.

☝🏽 this is based on my current understanding of the situation, without having done any further research.

File: pkg/helm/charts/awslbc/awslbc.go
Comment (paulbes) [reply pkg/helm/charts/awslbc/awslbc.go:39]: We are doing this many places already 😭 we should fix this, but perhaps as its own PR, so we can fix it at all places?

**Comment by paulbes** [comment]
@yngvark short answer: they don't. However, you are touching on something very interesting:

- When do we start introducing upgrades?
- To properly manage expectations, what should we communicate to our users?

I think this is something we need to discuss within the team, and come to some common understanding about?

File: pkg/api/core/service_serviceaccount.go
Comment (deifyed) [reply pkg/api/core/service_serviceaccount.go:186]: Ah. I thought clusterconfig was our code. My bad

File: pkg/clusterconfig/clusterconfig.go
Comment (deifyed) [reply pkg/clusterconfig/clusterconfig.go:230]: Might be. The function name makes me think it actually creates a service account. 

In any case, minor thing

File: pkg/helm/charts/awslbc/awslbc.go
Comment (deifyed) [reply pkg/helm/charts/awslbc/awslbc.go:36]: If its like how the nginx ingress controller + certbot stack works, it will terminate TLS in the ingress. In any case, its a nice improvement. More in line with the standard way of acquiring certificates with K8s. We can discuss this later

**Review by deifyed — APPROVED** [review]

**Comment by paulbes** [comment]
This PR might be left hanging for a long time if we wait for the addons permissions boundary problem to be fixed in a released version of eksctl.

Downgrading to the last stable version (https://github.com/weaveworks/eksctl/releases/tag/0.32.0) without addons while we wait for a fixed release.

**Pull request merged**
